File: App/Zup/Utility.cs
using System.Globalization;

namespace Zup;

public static class Utility
{
    public static int GetWeekNumber(DateTime date)
    {
        var ci = CultureInfo.CurrentCulture;

        return ci.Calendar.GetWeekOfYear(date, ci.DateTimeFormat.CalendarWeekRule, ci.DateTimeFormat.FirstDayOfWeek);
    }

    public static IEnumerable<WeekData> GetWeekData(int year)
    {
        var data = new Dictionary<int, WeekData>();
        var from = new DateTime(year, 1, 1);
        var to = new DateTime(year, 12, 31);

        for (var i = from; i <= to; i = i.AddDays(1))
        {
            var wn = GetWeekNumber(i);

            if (data.ContainsKey(wn))
            {
                continue;
            }

            switch (i.DayOfWeek)
            {
                case DayOfWeek.Sunday:
                    data[wn] = new WeekData
                    {
                        WeekNumber = wn,
                        Start = i,
                        End = i.AddDays(7).AddSeconds(-1)
                    };
                    break;
                case DayOfWeek.Monday:
                    data[wn] = new WeekData
                    {
                        WeekNumber = wn,
                        Start = i.AddDays(-1),
                        End = i.AddDays(6).AddSeconds(-1)
                    };
                    break;
                case DayOfWeek.Tuesday:
                    data[wn] = new WeekData
                    {
                        WeekNumber = wn,
                        Start = i.AddDays(-2),
                        End = i.AddDays(5).AddSeconds(-1)
                    };
                    break;
                case DayOfWeek.Wednesday:
                    data[wn] = new WeekData
                    {
                        WeekNumber = wn,
                        Start = i.AddDays(-3),
                        End = i.AddDays(4).AddSeconds(-1)
                    };
                    break;
                case DayOfWeek.Thursday:
                    data[wn] = new WeekData
                    {
                        WeekNumber = wn,
                        Start = i.AddDays(-4),
                        End = i.AddDays(3).AddSeconds(-1)
                    };
                    break;
                case DayOfWeek.Friday:
                    data[wn] = new WeekData
                    {
                        WeekNumber = wn,
                        Start = i.AddDays(-5),
                        End = i.AddDays(2).AddSeconds(-1)
                    };
                    break;
                case DayOfWeek.Saturday:
                    data[wn] = new WeekData
                    {
                        WeekNumber = wn,
                        Start = i.AddDays(-6),
                        End = i.AddDays(1).AddSeconds(-1)
                    };
                    break;
            }
        }

        return data.Select(a => a.Value)
            .ToArray();
    }

    public static DayOfWeek? GetDayOfWeek(DateTime? startedOn, TimeSpan dayStart, TimeSpan dayEnd)
    {
        if (startedOn == null)
        {
            return null;
        }

        var dt = new DateTime(startedOn.Value.Year, startedOn.Value.Month, startedOn.Value.Day);

        var ds1 = GetDayShift(dt.AddHours(-24), dayStart, dayEnd);
        var ds2 = GetDayShift(dt, dayStart, dayEnd);
        var ds3 = GetDayShift(dt.AddHours(24), dayStart, dayEnd);

        if (ds1.start <= startedOn.Value && startedOn.Value <= ds1.end)
        {
            return ds1.passedDT.DayOfWeek;
        }
        else if (ds2.start <= startedOn.Value && startedOn.Value <= ds2.end)
        {
            return ds2.passedDT.DayOfWeek;
        }
        else if (ds3.start <= startedOn.Value && startedOn.Value <= ds3.end)
        {
            return ds3.passedDT.DayOfWeek;
        }

        return null;
    }

    public static (DateTime start, DateTime end, DateTime passedDT) GetDayShift(DateTime date, TimeSpan dayStart, TimeSpan dayEnd)
    {
        var s = date;
        s = s.AddHours(dayStart.Hours - 7);
        s = s.AddMinutes(dayStart.Minutes);

        var e = date;
        e = e.AddHours(dayEnd.Hours + 8);
        e = e.AddMinutes(dayEnd.Minutes);
        e = e.AddSeconds(-1);

        e = e.AddHours(24);

        return (s, e, date);
    }
}

File: App/electron/services/utility.ts
// Port of Utility.cs

export interface WeekData {
	weekNumber: number;
	start: string; // ISO date string
	end: string; // ISO date string
	label: string;
}

export function getWeekNumber(date: Date): number {
	const d = new Date(Date.UTC(date.getFullYear(), date.getMonth(), date.getDate()));
	d.setUTCDate(d.getUTCDate() + 4 - (d.getUTCDay() || 7));
	const yearStart = new Date(Date.UTC(d.getUTCFullYear(), 0, 1));
	return Math.ceil(((d.getTime() - yearStart.getTime()) / 86400000 + 1) / 7);
}

export function getWeekData(year: number): WeekData[] {
	const data = new Map<number, WeekData>();
	const from = new Date(year, 0, 1);
	const to = new Date(year, 11, 31);

	for (let d = new Date(from); d <= to; d.setDate(d.getDate() + 1)) {
		const wn = getWeekNumber(d);
		if (data.has(wn)) continue;

		const day = d.getDay(); // 0=Sun, 1=Mon, ...
		const offset = day === 0 ? 0 : -day;
		const weekStart = new Date(d);
		weekStart.setDate(d.getDate() + offset);

		const weekEnd = new Date(weekStart);
		weekEnd.setDate(weekStart.getDate() + 6);
		weekEnd.setHours(23, 59, 59, 999);

		const fmt = (dt: Date) =>
			`${String(dt.getMonth() + 1).padStart(2, '0')}/${String(dt.getDate()).padStart(2, '0')}/${String(dt.getFullYear()).slice(2)}`;
		data.set(wn, {
			weekNumber: wn,
			start: weekStart.toISOString(),
			end: weekEnd.toISOString(),
			label: `${fmt(weekStart)} - ${fmt(weekEnd)}`,
		});
	}

	return Array.from(data.values()).sort((a, b) => a.weekNumber - b.weekNumber);
}

export function getDayShift(date: Date, dayStartStr: string, dayEndStr: string): { start: Date; end: Date } {
	const [startH, startM] = dayStartStr.split(':').map(Number);
	const [endH, endM] = dayEndStr.split(':').map(Number);

	const start = new Date(date);
	start.setHours(startH, startM, 0, 0);

	const end = new Date(date);
	end.setDate(end.getDate() + 1);
	end.setHours(endH, endM, 59, 999);

	return { start, end };
}

export function getDayOfWeek(startedOn: string | null, dayStart: string, dayEnd: string): string | null {
	if (!startedOn) return null;

	const dt = new Date(startedOn);
	const days = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];

	for (let offset = -1; offset <= 1; offset++) {
		const base = new Date(dt);
		base.setDate(base.getDate() + offset);
		base.setHours(0, 0, 0, 0);

		const shift = getDayShift(base, dayStart, dayEnd);
		if (dt >= shift.start && dt <= shift.end) {
			return days[base.getDay()];
		}
	}

	return null;
}
